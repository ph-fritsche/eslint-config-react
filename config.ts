import {createRequire} from 'node:module'
import process from 'node:process'

import eslint from '@eslint/js'
import TsParser from '@typescript-eslint/parser'
import JsxA11y from 'eslint-plugin-jsx-a11y'
import TsLint from '@typescript-eslint/eslint-plugin'
import Jest from 'eslint-plugin-jest'
import JestDom from 'eslint-plugin-jest-dom'
import React from 'eslint-plugin-react'
import ReactHooks from 'eslint-plugin-react-hooks'
import TestingLibrary from 'eslint-plugin-testing-library'
import globals from 'globals'
import {Linter} from 'eslint'
import Stylistic from '@stylistic/eslint-plugin'

const require = createRequire(import.meta.url)
function moduleExists(moduleName: string) {
    try {
        require.resolve(moduleName)
        return true
    } catch {
        return false
    }
}

/**
 * @internal
 */
export const filePatterns = {
    jsFiles: ['**/*.{js,jsx,mjs,cjs}'],
    tsFiles: ['**/*.{ts,tsx,mts,cts}'],
    jtsxFiles: ['**/*.[jt]sx'],
    testFiles: [
        ['{test,tests}/**', '**/*.{js,jsx,mjs,cjs,ts,tsx,mts,cts}'],
        ['**/__tests__/**', '**/*.{js,jsx,mjs,cjs,ts,tsx,mts,cts}'],
        '**/*.{test,spec}.{js,jsx,mjs,cjs,ts,tsx,mts,cts}',
    ],
    storyFiles: ['**/*.stories.{js,jsx,mjs,cjs,ts,tsx,mts,cts}'],
}

const config: Linter.Config[] = [
    {
        files: [...filePatterns.jsFiles],
    },
    {
        files: [...filePatterns.jtsxFiles],
        languageOptions: {
            parserOptions: {
                ecmaFeatures: {
                    jsx: true,
                },
            },
        },
    },
    eslint.configs.recommended,
]

// Typescript
if (moduleExists('typescript')) {
    config.push(
        {
            files: [...filePatterns.tsFiles],
            languageOptions: {
                parser: TsParser,
                parserOptions: {
                    projectService: true,
                },
            },
        },
        {
            files: [...filePatterns.tsFiles],
            plugins: {
                // eslint-disable-next-line @typescript-eslint/ban-ts-comment
                // @ts-ignore
                '@typescript-eslint': TsLint,
            },
            rules: {
                ...TsLint.configs.recommended.rules,

                ...TsLint.configs['recommended-requiring-type-checking'].rules,

                // this rule causes an error with the new eslint config format
                '@typescript-eslint/require-await': 'off',
            },
        },
        {
            files: [...filePatterns.tsFiles],
            rules: {
                'no-undef': 0,
                'no-redeclare': 0,
                'no-dupe-class-members': 0,
                '@typescript-eslint/no-floating-promises': [2, {
                    ignoreVoid: true,
                }],
                '@typescript-eslint/no-unsafe-argument': 1,
                '@typescript-eslint/no-unsafe-assignment': 1,
                '@typescript-eslint/no-unsafe-call': 1,
                '@typescript-eslint/no-unsafe-member-access': 1,
                '@typescript-eslint/no-unsafe-return': 1,
                '@typescript-eslint/restrict-template-expressions': [2, {
                    allowNumber: true,
                }],
            },
        },
    )
}

// Jest
if (moduleExists('jest')) {
    config.push(
        {
            files: [...filePatterns.testFiles],
            plugins: {
                'jest': Jest,
            },
            rules: Jest.configs.recommended.rules,
            languageOptions: {
                globals: globals.jest,
            },
        },
        {
            files: [...filePatterns.testFiles],
            plugins: {
                'jest-dom': JestDom,
            },
            rules: JestDom.configs.recommended.rules,
        },
    )
}

// React
if (moduleExists('react')) {
    config.push(
        {
            plugins: {
                'react': React,
            },
            rules: React.configs.recommended.rules,
            settings: {
                react: {
                    version: 'detect',
                },
            },
        },
        {
            plugins: {
                // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment
                'react-hooks': ReactHooks,
            },
            // eslint-disable-next-line @typescript-eslint/no-unsafe-assignment, @typescript-eslint/no-unsafe-member-access
            rules: ReactHooks.configs.recommended.rules,
        },
        {
            files: [...filePatterns.testFiles, ...filePatterns.storyFiles],
            rules: {
                'react/prop-types': 0,
            },
        },
    )
}

// TestingLib
if (moduleExists('react')) {
    config.push(
        {
            plugins: {
                'testing-library': TestingLibrary,
            },
            rules: TestingLibrary.configs.react.rules,
        },
    )
} else {
    config.push(
        {
            plugins: {
                'testing-library': TestingLibrary,
            },
            rules: TestingLibrary.configs.dom.rules,
        },
    )
}

// A11y
config.push(
    {
        files: [...filePatterns.jtsxFiles],
        plugins: {
            'jsx-a11y': JsxA11y,
        },
        rules: JsxA11y.configs.recommended.rules,
    },
)

// Stylistic
config.push(
    Stylistic.configs['disable-legacy'],
    Stylistic.configs['recommended-flat'],
    {
        rules: {
            '@stylistic/arrow-parens': 0,
            '@stylistic/brace-style': [2, '1tbs'],
            '@stylistic/indent': [2, 4],
            '@stylistic/object-curly-spacing': [1, 'never'],
            '@stylistic/object-curly-newline': [1, {
                multiline: true,
                consistent: true,
            }],
            '@stylistic/quote-props': [2, 'consistent'],
            '@stylistic/quotes': [2, 'single', {
                avoidEscape: true,
                allowTemplateLiterals: true,
            }],
            '@stylistic/semi': [2, 'never', {
                beforeStatementContinuationChars: 'always',
            }],
            '@stylistic/semi-style': [2, 'first'],
        },
    },
)

if (process.env.NODE_ENV === 'development') {
    config.push(
        {
            rules: {
                '@stylistic/indent': 1,
                'no-unreachable': 1,
            },
        },
        {
            files: [...filePatterns.jsFiles],
            rules: {
                'no-unused-vars': 1,
            },
        },
        {
            files: [...filePatterns.tsFiles],
            rules: {
                '@typescript-eslint/no-unused-vars': 1,
            },
        },
    )
}

config.push(
    {
        ignores: [
            'build/**',
            'coverage/**',
            'dist/**',
            'node_modules/**',
            'var/**',
            '**/__snapshots__/',
            '**/*.d.ts',
        ],
    },
)

export default config
