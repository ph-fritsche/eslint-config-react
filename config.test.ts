import {spawn} from 'node:child_process'
import {readdir} from 'node:fs/promises'
import {Buffer} from 'node:buffer'
import {cwd} from 'node:process'

function lint(
    file: string,
) {
    const child = spawn('node', [
        'node_modules/.bin/eslint',
        '-c',
        'config.ts',
        file,
    ], {
        stdio: ['pipe', 'pipe', 'pipe'],
        shell: true,
        env: {
            CI: 'true',
        },
    })

    const result = new Promise((res, rej) => {
        const outPromise = content(child.stdout)
        const errPromise = content(child.stderr)
        child.on('error', (e) => rej(e))
        child.on('exit', (code) => {
            void Promise.allSettled([outPromise, errPromise]).then(([out, err]) => {
                res({
                    code,
                    out: out.status === 'fulfilled' ? normalizeOutput(out.value.toString()) : String(out.reason),
                    err: err.status === 'fulfilled' ? err.value.toString() : String(err.reason),
                })
            })
        })
    })

    const kill = () => new Promise<void>((res) => {
        if (child.exitCode === null) {
            child.on('close', () => res())
            child.kill()
        } else {
            res()
        }
    })

    child.stdin.end()

    return {child, result, kill}
}

function content(
    stream: NodeJS.ReadableStream,
) {
    return new Promise<Buffer>((res, rej) => {
        const chunks: Buffer[] = []
        stream.on('data', (c: Buffer) => chunks.push(c))
        stream.on('error', rej)
        stream.on('end', () => res(Buffer.concat(chunks)))
    })
}

function normalizeOutput(
    output: string,
) {
    return output.replace(cwd() + '/', '')
}

const cleanup: Array<() => Promise<void>> = []
afterEach(async () => {
    const c = cleanup
    cleanup.splice(0)
    for (const f of c) {
        await f()
    }
})

for (const f of await readdir('examples')) {
    if (f === '.' || f === '..') {
        continue
    }

    test(`lint ${f}`, async () => {
        const {result, kill} = lint(`examples/${f}`)
        cleanup.push(kill)

        await expect(result).resolves.toMatchSnapshot()
    }, 30_000)
}

import {ConfigArray} from '@eslint/config-array'
import {filePatterns} from './config'

describe('match files', () => {
    const matches: {
        [k: string]: {
            [k in keyof typeof filePatterns]?: boolean
        }
    } = {
        '/example/foo.js': {jsFiles: true},
        '/example/foo.jsx': {jsFiles: true, jtsxFiles: true},
        '/example/foo.mjs': {jsFiles: true},
        '/example/foo.cjs': {jsFiles: true},
        '/example/foo.ts': {tsFiles: true},
        '/example/foo.tsx': {tsFiles: true, jtsxFiles: true},
        '/example/foo.mts': {tsFiles: true},
        '/example/foo.cts': {tsFiles: true},
        '/example/foo.html': {},
        '/example/foo/bar.js': {jsFiles: true},
        '/example/foo/bar.jsx': {jsFiles: true, jtsxFiles: true},
        '/example/foo/bar.mjs': {jsFiles: true},
        '/example/foo/bar.cjs': {jsFiles: true},
        '/example/foo/bar.ts': {tsFiles: true},
        '/example/foo/bar.tsx': {tsFiles: true, jtsxFiles: true},
        '/example/foo/bar.mts': {tsFiles: true},
        '/example/foo/bar.cts': {tsFiles: true},
        '/example/foo/bar.html': {},
        '/example/test/bar.js': {jsFiles: true, testFiles: true},
        '/example/test/bar.jsx': {jsFiles: true, jtsxFiles: true, testFiles: true},
        '/example/test/bar.mjs': {jsFiles: true, testFiles: true},
        '/example/test/bar.cjs': {jsFiles: true, testFiles: true},
        '/example/test/bar.ts': {tsFiles: true, testFiles: true},
        '/example/test/bar.tsx': {tsFiles: true, jtsxFiles: true, testFiles: true},
        '/example/test/bar.mts': {tsFiles: true, testFiles: true},
        '/example/test/bar.cts': {tsFiles: true, testFiles: true},
        '/example/test/bar.html': {},
        '/example/tests/bar.js': {jsFiles: true, testFiles: true},
        '/example/tests/bar.jsx': {jsFiles: true, jtsxFiles: true, testFiles: true},
        '/example/tests/bar.mjs': {jsFiles: true, testFiles: true},
        '/example/tests/bar.cjs': {jsFiles: true, testFiles: true},
        '/example/tests/bar.ts': {tsFiles: true, testFiles: true},
        '/example/tests/bar.tsx': {tsFiles: true, jtsxFiles: true, testFiles: true},
        '/example/tests/bar.mts': {tsFiles: true, testFiles: true},
        '/example/tests/bar.cts': {tsFiles: true, testFiles: true},
        '/example/tests/bar.html': {},
        '/example/foo/__tests__/bar.js': {jsFiles: true, testFiles: true},
        '/example/foo/__tests__/bar.jsx': {jsFiles: true, jtsxFiles: true, testFiles: true},
        '/example/foo/__tests__/bar.mjs': {jsFiles: true, testFiles: true},
        '/example/foo/__tests__/bar.cjs': {jsFiles: true, testFiles: true},
        '/example/foo/__tests__/bar.ts': {tsFiles: true, testFiles: true},
        '/example/foo/__tests__/bar.tsx': {tsFiles: true, jtsxFiles: true, testFiles: true},
        '/example/foo/__tests__/bar.mts': {tsFiles: true, testFiles: true},
        '/example/foo/__tests__/bar.cts': {tsFiles: true, testFiles: true},
        '/example/foo/__tests__/bar.html': {},
        '/example/foo/bar.stories.js': {jsFiles: true, storyFiles: true},
        '/example/foo/bar.stories.jsx': {jsFiles: true, jtsxFiles: true, storyFiles: true},
        '/example/foo/bar.stories.mjs': {jsFiles: true, storyFiles: true},
        '/example/foo/bar.stories.cjs': {jsFiles: true, storyFiles: true},
        '/example/foo/bar.stories.ts': {tsFiles: true, storyFiles: true},
        '/example/foo/bar.stories.tsx': {tsFiles: true, jtsxFiles: true, storyFiles: true},
        '/example/foo/bar.stories.mts': {tsFiles: true, storyFiles: true},
        '/example/foo/bar.stories.cts': {tsFiles: true, storyFiles: true},
        '/example/foo/bar.stories.html': {},
    }

    for (const patternName of Object.keys(filePatterns)) {
        test(patternName, () => {
            const config = new ConfigArray([
                {files: filePatterns[patternName] as string[]},
            ], {
                normalized: true,
                basePath: '/example',
            })

            const actual = {}
            const expected = {}

            for (const fileName of Object.keys(matches)) {
                actual[fileName] = config.getConfig(fileName) !== undefined
                expected[fileName] = !!matches[fileName][patternName]
            }

            expect(actual).toEqual(expected)
        })
    }
})
