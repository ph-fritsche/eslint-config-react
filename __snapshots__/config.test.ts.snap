// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`lint javascript.js 1`] = `
{
  "code": 1,
  "err": "",
  "out": "
examples/javascript.js
  2:10  error  'foo' is defined but never used               no-unused-vars
  4:1   error  Expected indentation of 4 spaces but found 8  @stylistic/indent

✖ 2 problems (2 errors, 0 warnings)
  1 error and 0 warnings potentially fixable with the \`--fix\` option.

",
}
`;

exports[`lint javascriptxml.jsx 1`] = `
{
  "code": 1,
  "err": "",
  "out": "
examples/javascriptxml.jsx
  4:10  error  'foo' is defined but never used                                                                                                                                                                    no-unused-vars
  6:12  error  Avoid non-native interactive elements. If using native HTML is not possible, add an appropriate role and support for tabbing, mouse, keyboard, and touch inputs to an interactive content element  jsx-a11y/no-static-element-interactions

✖ 2 problems (2 errors, 0 warnings)

",
}
`;

exports[`lint tsconfig.json 1`] = `
{
  "code": 0,
  "err": "",
  "out": "
examples/tsconfig.json
  0:0  warning  File ignored because no matching configuration was supplied

✖ 1 problem (0 errors, 1 warning)

",
}
`;

exports[`lint typescript.ts 1`] = `
{
  "code": 1,
  "err": "",
  "out": "
examples/typescript.ts
  2:10  error  'foo' is defined but never used                                                                                                                                     @typescript-eslint/no-unused-vars
  4:5   error  Promises must be awaited, end with a call to .catch, end with a call to .then with a rejection handler or be explicitly marked as ignored with the \`void\` operator  @typescript-eslint/no-floating-promises

✖ 2 problems (2 errors, 0 warnings)

",
}
`;

exports[`lint typescriptxml.tsx 1`] = `
{
  "code": 1,
  "err": "",
  "out": "
examples/typescriptxml.tsx
  4:10  error  'foo' is defined but never used                                                                                                                                                                    @typescript-eslint/no-unused-vars
  6:12  error  Avoid non-native interactive elements. If using native HTML is not possible, add an appropriate role and support for tabbing, mouse, keyboard, and touch inputs to an interactive content element  jsx-a11y/no-static-element-interactions

✖ 2 problems (2 errors, 0 warnings)

",
}
`;
